### PR TITLE
Make HTMLProps generic; add Intrinsic Attribute types

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
This adds two new types to the JSX namespace that will improve support for React in TypeScript 1.8 (see Microsoft/TypeScript#5478)

It also makes HTMLProps generic and updates all the built-in elements to be strongly typed to their respective DOM elements.

### CODE REVIEW

File: react/react.d.ts [hunk 1]
```diff
@@ -323,7 +323,7 @@ declare namespace __React {
         ref?: string | ((component: T) => any);
     }
 
-    interface HTMLProps extends HTMLAttributes, Props<HTMLElement> {
+    interface HTMLProps<T> extends HTMLAttributes, Props<T> {
```
Comment (vsiao) [inline react/react.d.ts]: `<T extends HTMLElement>`?

**Comment by vvakame** [comment]
@pspeter3 @jbrantly ?

**Comment by jbrantly** [comment]
:+1: from me

**Comment by pspeter3** [comment]
I agree with @vsiao's comment but otherwise looks good.

**Comment by vvakame** [comment]
thanks! I'm wait for response from @RyanCavanaugh

**Pull request merged**